modules/caddyhttp/push/handler.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2015 Matthew Holt and The Caddy Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package push
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/modules/caddyhttp"
+	"github.com/caddyserver/caddy/v2/modules/caddyhttp/headers"
+	"go.uber.org/zap"
+)
+
+func init() {
+	caddy.RegisterModule(Handler{})
+}
+
+// Handler is a middleware for manipulating the request body.
+type Handler struct {
+	Resources []Resource    `json:"resources,omitempty"`
+	Headers   *HeaderConfig `json:"headers,omitempty"`
+
+	logger *zap.Logger
+}
+
+// CaddyModule returns the Caddy module information.
+func (Handler) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  "http.handlers.push",
+		New: func() caddy.Module { return new(Handler) },
+	}
+}
+
+// Provision sets up h.
+func (h *Handler) Provision(ctx caddy.Context) error {
+	h.logger = ctx.Logger(h)
+	if h.Headers != nil {
+		err := h.Headers.Provision(ctx)
+		if err != nil {
+			return fmt.Errorf("provisioning header operations: %v", err)
+		}
+	}
+	return nil
+}
+
+func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request, next caddyhttp.Handler) error {
+	pusher, ok := w.(http.Pusher)
+	if !ok {
+		return next.ServeHTTP(w, r)
+	}
+
+	// short-circuit recursive pushes
+	if _, ok := r.Header[pushHeader]; ok {
+		return next.ServeHTTP(w, r)
+	}
+
```

Comment:
This means something like index.html > main.css > img.png would not be possible to push on a request to index.html?

Reply:
I don't really know what you mean by that. Can you explain the "something like" part more?

Reply:
If index.html is configured to push main.css and main.css is configured to push img.png, and a user requests index.html, Caddy will only push main.css, because the synthetic request to main.css contains the header and suppresses further pushes, no?

Reply:
Correct

Reply:
It's possible to "workaround" this behavior by computing recursively all resources to push, send the push promises and the push responses in background, but only start pushing the content of the explicit response when all resources to push has been resolved (and all push promises have been sent).

It's what we do in the Vulcain Gateway Server, it should be relatively easy to port the code.

Reply:
This is only possible if pushes are configured statically.

Reply:
It can work with dynamic pushes too (Vulcain supports dynamic client-requested pushes).

Reply:
Is that a merge blocker? Can we implement that later?

Reply:
We definitely can implement this later (and I can work on this, I've already done it once, I can do it again 😄).

Reply:
Great! I'll merge this in a moment then, and we can improve on it later. Thanks for the feedback, and I'd be very happy to work with you on improvements and/or review a PR for it.